aten/src/ATen/native/cuda/TensorFactories.cu

```diff
@@ -26,4 +36,57 @@ Tensor& eye_out_cuda(Tensor& result, int64_t n, int64_t m) {
   return result;
 }
 
+Tensor& randperm_out_cuda(Tensor& result, int64_t n, Generator* generator) {
+  if (n < 0) {
+    std::ostringstream oss;
+    oss << "n must be non-negative, got " << n;
+    throw std::runtime_error(oss.str());
+  }
+
+  if (n > 0) {
+    AT_DISPATCH_ALL_TYPES_AND_HALF(
+      result.type(), "randperm_out_cuda", [&] {
+        AT_CHECK(Scalar(n).to<scalar_t>(),
+          "n is too large for result tensor type: '", result.type().toString(), "'");
+      }
+    );
+  }
+
+  result.resize_({n});
+
+  if (result.type().scalarType() == at::ScalarType::Half) {
+    auto result_float = CUDA(kFloat).tensor({n});
+    result.copy_(randperm_out_cuda(result_float, n, generator));
+  } else {
+    if (n < 30000) {  // For small inputs, we offload it to CPU instead.
+      auto result_cpu = result.type().toBackend(kCPU).tensor({n});
+      randperm_out(result_cpu, n, generator);
+      result.copy_(result_cpu);
+    } else {
+      // Generate random values for the keys array
+      AT_DISPATCH_ALL_TYPES(
+        result.type(), "randperm_out_cuda", [&] {
+          using cuda_scalar_t = cuda::into_type<scalar_t>;
+
+          auto keys = result.type().tensor(result.sizes()).random_(generator);
```

Comment:
Did you try `keys.sort()`? The sorting in PyTorch (written by Jeff) is very fast for small tensors. IIRC, it uses thrust internally for large tensors.

Much better small tensor perf:
```
%timeit x = torch.cuda.LongTensor(10); x.random_(); x.sort(); torch.cuda.synchronize()
10000 loops, best of 3: 48.8 µs per loop
```

Same large tensor perf:
```
%timeit x = torch.cuda.LongTensor(10000000); x.random_(); x.sort(); torch.cuda.synchronize()
10 loops, best of 3: 58.4 ms per loop
```